File: src/agentsfleetd/cron/Store.zig
//! Postgres persistence for hosted schedules.
//!
//! The store retains only the pool pointer. Returned rows are allocated with
//! the allocator passed to each call and must be released with `Schedule.deinit`.

const Store = @This();

const std = @import("std");
const pg = @import("pg");
const logging = @import("log");

const PgQuery = @import("../db/pg_query.zig").PgQuery;
const ec = @import("../errors/error_registry.zig");
const model = @import("model.zig");
const sql = @import("sql.zig");

const log = logging.scoped(.cron_store);

pool: *pg.Pool,

pub const CreateOutcome = union(enum) {
    created: model.Schedule,
    fleet_not_found,
    cap_reached,
    source_conflict,
};

pub const ClaimOutcome = union(enum) {
    claimed: model.Schedule,
    busy,
    not_found,
};

pub fn init(pool: *pg.Pool) Store {
    return .{ .pool = pool };
}

pub fn create(
    self: Store,
    alloc: std.mem.Allocator,
    input: model.CreateInput,
    schedule_id: []const u8,
    lease_token: []const u8,
    now_ms: i64,
    lease_until_ms: i64,
) !CreateOutcome {
    const conn = try self.pool.acquire();
    defer self.pool.release(conn);

    _ = try conn.exec("BEGIN", .{});
    var tx_open = true;
    defer if (tx_open) rollbackLogged(conn);

    const count = try lockedScheduleCount(conn, input.fleet_id) orelse {
        try rollback(conn, &tx_open);
        return .{ .fleet_not_found = {} };
    };
    if (try sourceKeyExists(conn, input.fleet_id, input.source_key)) {
        try rollback(conn, &tx_open);
        return .{ .source_conflict = {} };
    }
    if (count >= model.MAX_SCHEDULES_PER_FLEET) {
        try rollback(conn, &tx_open);
        return .{ .cap_reached = {} };
    }

    var schedule = try insertRow(conn, alloc, input, schedule_id, lease_token, now_ms, lease_until_ms);
    errdefer schedule.deinit(alloc);
    _ = try conn.exec("COMMIT", .{});
    tx_open = false;
    return .{ .created = schedule };
}

pub fn get(self: Store, alloc: std.mem.Allocator, fleet_id: []const u8, schedule_id: []const u8) !?model.Schedule {
    const conn = try self.pool.acquire();
    defer self.pool.release(conn);
    var query = PgQuery.from(try conn.query(sql.SELECT_ONE, .{ schedule_id, fleet_id }));
    defer query.deinit();
    const row = (try query.next()) orelse return null;
    return try rowToSchedule(alloc, row);
}

pub fn list(self: Store, alloc: std.mem.Allocator, fleet_id: []const u8) ![]model.Schedule {
    const conn = try self.pool.acquire();
    defer self.pool.release(conn);
    var query = PgQuery.from(try conn.query(sql.LIST_FOR_FLEET, .{fleet_id}));
    defer query.deinit();

    var rows: std.ArrayList(model.Schedule) = .empty;
    errdefer {
        for (rows.items) |*row| row.deinit(alloc);
        rows.deinit(alloc);
    }
    while (try query.next()) |row| {
        var schedule = try rowToSchedule(alloc, row);
        errdefer schedule.deinit(alloc);
        try rows.append(alloc, schedule);
    }
    return rows.toOwnedSlice(alloc);
}

pub fn claimMutation(self: Store, alloc: std.mem.Allocator, input: model.MutationInput) !ClaimOutcome {
    const conn = try self.pool.acquire();
    defer self.pool.release(conn);
    {
        var query = PgQuery.from(try conn.query(sql.CLAIM_MUTATION, .{
            input.schedule_id,
            input.fleet_id,
            input.cron,
            input.timezone,
            input.message,
            input.desired_status.toSlice(),
            model.SyncStatus.syncing.toSlice(),
            input.lease_token,
            input.lease_until_ms,
            input.now_ms,
        }));
        defer query.deinit();
        if (try query.next()) |row| return .{ .claimed = try rowToSchedule(alloc, row) };
    }
    return if (try scheduleExists(conn, input.fleet_id, input.schedule_id))
        .{ .busy = {} }
    else
        .{ .not_found = {} };
}

pub fn finalizeSuccess(
    self: Store,
    alloc: std.mem.Allocator,
    schedule_id: []const u8,
    generation: i64,
    lease_token: []const u8,
    now_ms: i64,
) !?model.Schedule {
    return self.finalize(alloc, sql.FINALIZE_SUCCESS, .{ schedule_id, generation, lease_token, model.SyncStatus.synced.toSlice(), now_ms });
}

pub fn finalizeFailure(
    self: Store,
    alloc: std.mem.Allocator,
    schedule_id: []const u8,
    generation: i64,
    lease_token: []const u8,
    detail: []const u8,
    now_ms: i64,
) !?model.Schedule {
    return self.finalize(alloc, sql.FINALIZE_FAILURE, .{ schedule_id, generation, lease_token, model.SyncStatus.failed.toSlice(), detail, now_ms });
}

pub fn deleteClaimed(self: Store, schedule_id: []const u8, generation: i64, lease_token: []const u8) !bool {
    const conn = try self.pool.acquire();
    defer self.pool.release(conn);
    var query = PgQuery.from(try conn.query(sql.DELETE_CLAIMED, .{ schedule_id, generation, lease_token }));
    defer query.deinit();
    return (try query.next()) != null;
}

fn finalize(self: Store, alloc: std.mem.Allocator, statement: []const u8, args: anytype) !?model.Schedule {
    const conn = try self.pool.acquire();
    defer self.pool.release(conn);
    var query = PgQuery.from(try conn.query(statement, args));
    defer query.deinit();
    const row = (try query.next()) orelse return null;
    return try rowToSchedule(alloc, row);
}

fn lockedScheduleCount(conn: *pg.Conn, fleet_id: []const u8) !?usize {
    var query = PgQuery.from(try conn.query(sql.LOCK_FLEET_AND_COUNT, .{fleet_id}));
    defer query.deinit();
    const row = (try query.next()) orelse return null;
    return @intCast(try row.get(i64, 1));
}

fn sourceKeyExists(conn: *pg.Conn, fleet_id: []const u8, source_key: []const u8) !bool {
    var query = PgQuery.from(try conn.query(sql.SOURCE_KEY_EXISTS, .{ fleet_id, source_key }));
    defer query.deinit();
    return (try query.next()) != null;
}

fn scheduleExists(conn: *pg.Conn, fleet_id: []const u8, schedule_id: []const u8) !bool {
    var query = PgQuery.from(try conn.query(sql.EXISTS, .{ schedule_id, fleet_id }));
    defer query.deinit();
    return (try query.next()) != null;
}

fn insertRow(
    conn: *pg.Conn,
    alloc: std.mem.Allocator,
    input: model.CreateInput,
    schedule_id: []const u8,
    lease_token: []const u8,
    now_ms: i64,
    lease_until_ms: i64,
) !model.Schedule {
    var query = PgQuery.from(try conn.query(sql.INSERT, .{
        schedule_id,
        input.fleet_id,
        input.source.toSlice(),
        input.source_key,
        input.cron,
        input.timezone,
        input.message,
        model.DesiredStatus.active.toSlice(),
        model.SyncStatus.syncing.toSlice(),
        @as(i64, 1),
        lease_token,
        lease_until_ms,
        now_ms,
    }));
    defer query.deinit();
    const row = (try query.next()) orelse return error.InsertReturnedNoRow;
    return rowToSchedule(alloc, row);
}

fn rowToSchedule(alloc: std.mem.Allocator, row: pg.Row) !model.Schedule {
    const source = model.Source.fromSlice(try row.get([]const u8, 2)) orelse return error.InvalidScheduleRow;
    const desired = model.DesiredStatus.fromSlice(try row.get([]const u8, 7)) orelse return error.InvalidScheduleRow;
    const sync = model.SyncStatus.fromSlice(try row.get([]const u8, 8)) orelse return error.InvalidScheduleRow;
    const schedule_id = try alloc.dupe(u8, try row.get([]const u8, 0));
    errdefer alloc.free(schedule_id);
    const fleet_id = try alloc.dupe(u8, try row.get([]const u8, 1));
    errdefer alloc.free(fleet_id);
    const source_key = try alloc.dupe(u8, try row.get([]const u8, 3));
    errdefer alloc.free(source_key);
    const cron = try alloc.dupe(u8, try row.get([]const u8, 4));
    errdefer alloc.free(cron);
    const timezone = try alloc.dupe(u8, try row.get([]const u8, 5));
    errdefer alloc.free(timezone);
    const message = try alloc.dupe(u8, try row.get([]const u8, 6));
    errdefer alloc.free(message);
    const sync_token = try optionalDupe(alloc, try row.get(?[]const u8, 10));
    errdefer if (sync_token) |value| alloc.free(value);
    const last_error = try optionalDupe(alloc, try row.get(?[]const u8, 12));
    errdefer if (last_error) |value| alloc.free(value);
    return .{
        .schedule_id = schedule_id,
        .fleet_id = fleet_id,
        .source = source,
        .source_key = source_key,
        .cron = cron,
        .timezone = timezone,
        .message = message,
        .desired_status = desired,
        .sync_status = sync,
        .generation = try row.get(i64, 9),
        .sync_token = sync_token,
        .sync_lease_until = try row.get(?i64, 11),
        .last_error = last_error,
        .created_at = try row.get(i64, 13),
        .updated_at = try row.get(i64, 14),
    };
}

fn optionalDupe(alloc: std.mem.Allocator, value: ?[]const u8) !?[]const u8 {
    return if (value) |bytes| try alloc.dupe(u8, bytes) else null;
}

fn rollback(conn: *pg.Conn, tx_open: *bool) !void {
    try conn.rollback();
    tx_open.* = false;
}

fn rollbackLogged(conn: *pg.Conn) void {
    conn.rollback() catch |err| log.warn("schedule_store_rollback_failed", .{
        .error_code = ec.ERR_INTERNAL_OPERATION_FAILED,
        .err = @errorName(err),
    });
}

File: rustd/crates/afd_cron/src/store.rs
//! The schedule table's reads and writes, and the fence that serialises them.
//!
//! # Every mutation leaves the row CLAIMED
//!
//! A create, an edit and a `:sync` all end with `sync_status = syncing`, a
//! token, and a lease. That is not bookkeeping: the caller is expected to go and
//! talk to the external scheduler next, and the claim is what stops a second
//! caller doing the same thing at the same time. The pair to every claim is a
//! finalize, which releases the fence and records whether the push worked.
//!
//! A caller that claims and then dies leaves a lease that expires. The next
//! claim takes the row because the fence predicate admits a lease in the past —
//! which is why the lease exists rather than a plain "held" flag, and why a
//! crashed syncer costs a lease window rather than a stuck schedule.
//!
//! # Absent is `Ok(None)`, not an error
//!
//! A schedule id that names no row, one belonging to another fleet, and one
//! another syncer is holding are all `Ok(None)`. None of them is this daemon
//! failing; each is a different answer the caller renders. Only the datastore
//! refusing, or a row this build cannot read, is an [`Error`].

use afd_core::clock::UnixMillis;
use afd_core::id::{ENTROPY_LEN, Uuid7};
use afd_crypto::entropy::Entropy;
use afd_db::Db;
use sqlx::{Acquire as _, Row as _};

use crate::error::{self, COLUMN_DESIRED_STATUS, COLUMN_FLEET, COLUMN_WORKSPACE, Result};
use crate::model::{DesiredStatus, MAX_SCHEDULES_PER_FLEET, Schedule, Source, SyncStatus};
use crate::sql;

mod decode;
mod fence;

use self::decode::decode;

/// The context a failed read reports under.
const CONTEXT_READ: &str = "read a schedule";

/// The context a failed write reports under.
const CONTEXT_WRITE: &str = "write a schedule";

/// How long a syncer's hold on a row lasts before another may take it.
///
/// Long enough for an upstream round trip and its retries, short enough that a
/// syncer killed mid-push does not strand a schedule for a person watching the
/// list. `FireStore.zig` uses the same window.
pub const SYNC_LEASE_MS: i64 = 30_000;

/// Why a create was refused.
///
/// Refusals rather than errors, for the reason [`crate::error`] gives: an
/// operator hit a bound, and nothing in this daemon failed.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Refused {
    /// The fleet is not in the workspace the caller was proven in.
    ///
    /// Answered identically to a fleet that does not exist — telling them apart
    /// would confirm a fleet id across a workspace boundary.
    NoSuchFleet,
    /// The fleet already holds [`MAX_SCHEDULES_PER_FLEET`].
    TooMany,
    /// This fleet already registered that upstream key.
    DuplicateKey,
}

/// What a schedule is created from.
///
/// A struct rather than seven positional arguments because four of them are
/// `&str` a call site could transpose without the compiler noticing.
#[derive(Debug, Clone, Copy)]
pub struct NewSchedule<'a> {
    /// The fleet the schedule wakes.
    pub fleet: &'a Uuid7,
    /// Who asked for it.
    pub source: Source,
    /// The key the external scheduler will know it by.
    pub source_key: &'a str,
    /// The expression, already through [`crate::validate::cron`].
    pub cron: &'a str,
    /// The zone, already through [`crate::validate::timezone`].
    pub timezone: &'a str,
    /// The message, already through [`crate::validate::message`].
    pub message: &'a str,
}

/// What an edit changes, field by field.
///
/// Every field is optional and an absent one is left alone — the `COALESCE`
/// half of [`sql::CLAIM_MUTATION`]. A struct of `Option` rather than a full
/// replacement value, because a partial edit expressed as a whole row would
/// make every caller read before it writes and race its own read.
#[derive(Debug, Clone, Copy, Default)]
pub struct Change<'a> {
    /// A new expression, or the one already stored.
    pub cron: Option<&'a str>,
    /// A new zone, or the one already stored.
    pub timezone: Option<&'a str>,
    /// A new message, or the one already stored.
    pub message: Option<&'a str>,
    /// A new intent, or the one already stored.
    pub desired_status: Option<DesiredStatus>,
}

/// Where a fire is going, once a signed callback resolved to it.
#[derive(Debug, Clone)]
pub struct FireTarget {
    /// The fleet to wake.
    pub fleet: Uuid7,
    /// The workspace the event is stamped with.
    pub workspace: Uuid7,
    /// What the fleet is asked to do.
    pub message: String,
    /// What the operator wants this schedule to be doing.
    pub desired_status: DesiredStatus,
    /// What the fleet's own row says it is doing.
    pub fleet_status: String,
}

/// The schedules table, over an already-connected pool.
///
/// Cheap to clone: [`Db`] is a handle over a shared pool.
#[derive(Debug, Clone)]
pub struct Schedules {
    /// Where every statement here runs.
    database: Db,
    /// Where a new schedule's identifier comes from.
    ///
    /// Held rather than taken per call for the reason `afd_admin::Models` holds
    /// one: the source is chosen by the binary, and a suite drives the mocked
    /// one so a created row's identifier is a fact a test can name.
    entropy: Entropy,
}

impl Schedules {
    /// Binds the store to an already-connected pool.
    #[must_use]
    pub const fn new(database: Db, entropy: Entropy) -> Self {
        Self { database, entropy }
    }

    /// A new schedule's identifier, ordered by the instant it was minted.
    ///
    /// `afd_admin::Models::mint`'s shape: the instant leads so rows sort by
    /// creation without a second column, and the entropy is what makes two
    /// minted in the same millisecond distinct.
    fn mint(&self, now: UnixMillis) -> Result<Uuid7> {
        let mut bytes = [0_u8; ENTROPY_LEN];
        self.entropy.fill(&mut bytes)?;
        Ok(Uuid7::encode(now, bytes)?)
    }

    /// This fleet's schedules, oldest first.
    ///
    /// # Errors
    /// Reports a datastore that would not answer and a row this build cannot
    /// read.
    pub async fn list(&self, fleet: &Uuid7) -> Result<Vec<Schedule>> {
        let mut connection = self.database.acquire().await?;
        let rows = sqlx::query(sql::LIST_FOR_FLEET)
            .bind(fleet.as_str())
            .fetch_all(connection.as_mut())
            .await
            .map_err(error::query(CONTEXT_READ))?;

        rows.iter().map(decode).collect()
    }

    /// One schedule of this fleet's.
    ///
    /// # Errors
    /// As [`Self::list`].
    pub async fn one(&self, fleet: &Uuid7, schedule: &Uuid7) -> Result<Option<Schedule>> {
        let mut connection = self.database.acquire().await?;
        let row = sqlx::query(sql::SELECT_ONE)
            .bind(schedule.as_str())
            .bind(fleet.as_str())
            .fetch_optional(connection.as_mut())
            .await
            .map_err(error::query(CONTEXT_READ))?;

        row.as_ref().map(decode).transpose()
    }

    /// Creates a schedule, already claimed by the caller that will register it.
    ///
    /// The count, the duplicate check and the insert run inside ONE transaction
    /// that has taken the fleet's row — see [`sql::LOCK_FLEET`] on why a lock
    /// on the parent is what makes a count-then-insert atomic.
    ///
    /// # Errors
    /// Reports a datastore that would not answer, a statement that failed, and
    /// a row this build cannot read. A bound the operator hit is `Ok(Err(..))`.
    pub async fn create(
        &self,
        workspace: &Uuid7,
        new: NewSchedule<'_>,
        token: &Uuid7,
        now: UnixMillis,
    ) -> Result<core::result::Result<Schedule, Refused>> {
        let schedule_id = self.mint(now)?;
        let mut connection = self.database.acquire().await?;
        let mut transaction = connection
            .begin()
            .await
            .map_err(error::query(CONTEXT_WRITE))?;

        let in_workspace = sqlx::query(sql::FLEET_IN_WORKSPACE)
            .bind(new.fleet.as_str())
            .bind(workspace.as_str())
            .fetch_optional(transaction.as_mut())
            .await
            .map_err(error::query(CONTEXT_WRITE))?;
        if in_workspace.is_none() {
            return Ok(Err(Refused::NoSuchFleet));
        }

        let _locked = sqlx::query(sql::LOCK_FLEET)
            .bind(new.fleet.as_str())
            .fetch_optional(transaction.as_mut())
            .await
            .map_err(error::query(CONTEXT_WRITE))?;

        let count: i64 = sqlx::query(sql::COUNT_FOR_FLEET)
            .bind(new.fleet.as_str())
            .fetch_one(transaction.as_mut())
            .await
            .map_err(error::query(CONTEXT_WRITE))?
            .try_get(0)
            .map_err(error::query(CONTEXT_WRITE))?;
        if usize::try_from(count).unwrap_or(usize::MAX) >= MAX_SCHEDULES_PER_FLEET {
            return Ok(Err(Refused::TooMany));
        }

        let duplicate = sqlx::query(sql::SOURCE_KEY_EXISTS)
            .bind(new.fleet.as_str())
            .bind(new.source_key)
            .fetch_optional(transaction.as_mut())
            .await
            .map_err(error::query(CONTEXT_WRITE))?;
        if duplicate.is_some() {
            return Ok(Err(Refused::DuplicateKey));
        }

        let row = sqlx::query(sql::INSERT)
            .bind(schedule_id.as_str())
            .bind(new.fleet.as_str())
            .bind(new.source.as_str())
            .bind(new.source_key)
            .bind(new.cron)
            .bind(new.timezone)
            .bind(new.message)
            .bind(DesiredStatus::Active.as_str())
            .bind(SyncStatus::Syncing.as_str())
            // Generation ONE, never zero: the column's own CHECK refuses zero,
            // because "never synced" and "synced at generation zero" would
            // otherwise be the same state to a finalize.
            .bind(1_i64)
            .bind(token.as_str())
            .bind(now.as_millis() + SYNC_LEASE_MS)
            .bind(now.as_millis())
            .fetch_one(transaction.as_mut())
            .await
            .map_err(error::query(CONTEXT_WRITE))?;

        let created = decode(&row)?;
        transaction
            .commit()
            .await
            .map_err(error::query(CONTEXT_WRITE))?;
        Ok(Ok(created))
    }

    /// What a signed fire resolves to.
    ///
    /// `Ok(None)` for a schedule this daemon has no row for — a callback the
    /// external scheduler kept after the schedule was deleted. Dropped rather
    /// than refused: the sender is correctly configured and acting on what it
    /// was last told.
    ///
    /// # Errors
    /// As [`Self::list`].
    pub async fn fire_target(&self, schedule: &Uuid7) -> Result<Option<FireTarget>> {
        let mut connection = self.database.acquire().await?;
        let row = sqlx::query(sql::FIRE_TARGET)
            .bind(schedule.as_str())
            .fetch_optional(connection.as_mut())
            .await
            .map_err(error::query(CONTEXT_READ))?;

        let Some(row) = row else {
            return Ok(None);
        };

        let unreadable = error::query(CONTEXT_READ);
        let fleet: String = row.try_get(0).map_err(&unreadable)?;
        let workspace: String = row.try_get(1).map_err(&unreadable)?;
        let message: String = row.try_get(2).map_err(&unreadable)?;
        let desired: String = row.try_get(3).map_err(&unreadable)?;
        let fleet_status: String = row.try_get(4).map_err(&unreadable)?;

        Ok(Some(FireTarget {
            fleet: Uuid7::parse(&fleet).map_err(|_shape| error::row_unreadable(COLUMN_FLEET))?,
            workspace: Uuid7::parse(&workspace)
                .map_err(|_shape| error::row_unreadable(COLUMN_WORKSPACE))?,
            message,
            desired_status: DesiredStatus::parse(&desired)
                .ok_or_else(|| error::row_unreadable(COLUMN_DESIRED_STATUS))?,
            fleet_status,
        }))
    }
}
